### DISCUSSION TITLE
Git rename but GitHub shows it as added file and deleted file

### DISCUSSION THREAD

**Discussion by shirady** [community/community#43177] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Bug

### Body

Hi,
The issue in detail:
1. I moved the file into a new directory (I dragged the file to the new directory using VS Code UI).
2. Git recognized this change as a deleted file and a new file, I added both changes before I commit.
3. I run git status, and you can see that the file `test_ceph_s3.js`  was recognized now as renamed 
4. I commit and push to the [PR](https://github.com/noobaa/noobaa-core/pull/7142/files) - but GitHub shows `test_ceph_s3.js` as [added](https://github.com/noobaa/noobaa-core/pull/7142/files#diff-fc8a74e235c86104d11a194fd2c1e15ef4ec4a85ada0978c5d6ff9c9af5ad054) and [deleted](https://github.com/noobaa/noobaa-core/pull/7142/files#diff-f60177e44f3c2a5477dc0743932d8214af8a2e30c738e518e244c6baf1bae19b) file.

Note: this was not the only file that I moved and renamed in this PR, all the other file was recognized as renamed in Git and also in GitHub.
I read [here](https://github.com/orgs/community/discussions/8573) that I should have used git mv (which as you can read above I didn't use...), and maybe it would fix it, but I'm not sure it would work since Git recognized the change of the file as renamed.

Thanks,
Shira

**Comment by LuisContrerasGlz** [comment]
try using the git mv command to explicitly rename the file, rather than simply moving it in the file system. This will help Git track the file's history more accurately.

**Reply by shirady** [reply]
Hi @LuisContrerasGlz
Git tracked the file history, it's GitHub that didn't...

**Reply by LuisContrerasGlz** [reply]
Oh I see sorry about the confusion

**Reply by danon** [reply]
On my machine, it doesn't make a difference if it's `giv mv` or `mv`.

**Comment by airtower-luna** [comment]
Git is very good about detecting when files are renamed, and as you described it did for you. I assume this is simply a limitation of the GitHub UI.

**Reply by shirady** [reply]
I just wonder what will happen, since it's a shared project, and we use the feature of 'Blame' - in case it detects it as delete and add I'm loosing all the previous changes.

**Reply by airtower-luna** [reply]
The [`git blame` documentation](https://git-scm.com/docs/git-blame) says this:

> The origin of lines is automatically followed across whole-file renames (currently there is no option to turn the rename-following off). To follow lines moved from one file to another, or to follow lines that were copied and pasted from another file, etc., see the `-C` and `-M` options.

So it should just work, and if it doesn't it's definitely a bug that the devs should want to fix. :slightly_smiling_face:

**Comment by spenserblack** [comment]
Just curious, were you using Windows when you did this? Git on Windows can automatically convert line endings depending on settings, so perhaps the commit tracks different line endings than the original file had.

**Reply by shirady** [reply]
Hi,
I'm using Mac :)

**Reply by spenserblack** [reply]
Thanks for answering! Just double-checked in a codespace, and I'm 99% sure line endings aren't the issue after all.

Also, I used `git show -- src/test/system_tests` and searched for the file (`test_ceph_s3.js`) and it shows the file as deleted in `system_tests` and created in `ceph_s3_tests`, so it looks like this isn't a GitHub-specific issue :thinking:

**Reply by spenserblack** [reply]
After renaming, did you make any edits? I'm looking at `src/test/system_tests/test_ceph_s3.js` from 1 commit prior to your PR, and it has a massive array called `S3_CEPH_TEST_BLACKLIST` that `src/test/system_tests/ceph_s3_tests/test_ceph_s3.js` doesn't seem to have.

If a diff is too large, git will detect it as a delete+add instead of a move.

Edit:
I see now in your comment that you amended the commit. The rename from `git status` kind of shows what would happen if you made a *new* commit, but because you edited an existing commit git decided "hey, the paths are different, and the contents look really different, so they must be different files."

If you don't like this, the only option that I know of is to move and edit in 2 different commits.

Hope this helps!

**Reply by shirady** [reply]
Thank you!

I have a couple of questions:

1. Would you please attach a link about this?
 > If a diff is too large, git will detect it as a delete+add instead of a move.
2. Why does it matter to `git status` if the commit is new or just amended? doesn't the result should be the same...? (in the aspect of the files)

**Reply by spenserblack** [reply]
I'll admit I don't have any docs handy, as I'm mostly basing this off of behavior I've seen from usage. Might try to hunt down docs later, but basically (to the best of my knowledge):

When you ran git status, your pending change, renaming the file, was separate from your last change, editing the file. So `git status` said that you had a rename that was *staged.* And it would have stayed as a rename if it was a new commit. But amending the existing changes makes git take previously committed *and* staged changes into consideration when generating a new diff, not just the staged changes.

To kind of simulate what happened, try this:
```bash
# in some temp dir
git init
echo "hello, world!" > file
git add --all
git commit -m "init"
echo "goobye, world!" > file
git add --all
git status  # shows we've modified the file. Something would be really wrong if it didn't :)
git mv file message

git status
# so we've moved the file, so it should say it's moved, right?
# no, git actually considers it a different file, because we've changed a large
# portion (in this case, 100%) of the file!
```

And, just to further elaborate, if we did the edit and move in reverse order:

```bash
echo "hello, world!" > file
git add --all
git commit -m "init"
git mv file message
git status  # shows we've moved the file
echo "goodbye, world!" > message
git status  # we have a staged rename, and an unstaged change
git add --all

git status
# once again, it's now a delete+add, not a move
```

AFAIK `mv` isn't really a special command that tells tracks a move event specifically. It's just a handy shortcut to move the file, untrack the old file, and track the new file. Git really cares about the contents of the file, so if the contents are very different (percent of lines changed), git doesn't have enough info to know that it's actually the *same* file, just with a different path and different contents.

Amending the commit is effectively the same as if you did all those changes in the same commit the first time around (the first code example).

**Reply by shirady** [reply]
Thank you for the detailed examples! 
(I would just remove the exclamation mark "!" from the printing, so people could easily copy it).
I thought that maybe squashing the commits would help, but from [what I tried](https://github.com/shirady/learn-amend/commit/7ddb1fa9143073b2cf2016104b288aff4435ff96) on GitHub it still shows it as add/delete file.
Note: I added a link to a temporary repo, and it will be deleted in the future.

**Reply by mkasem06** [reply]
Hi there, 
Did u manage to find any solution?
I'm going through the same situation, I've a big file difference snd git recognizes it as a rename it's just when I push it, it appears as a new/del file on GitHub. I've tried everything possible (rebasing, 2 different commits), everything and it's still not working. Maybe there's a way to increase the diff that is needed to recognize it as a new/del file?

**Reply by shirady** [reply]
Hi @CoolGuyMK,
I didn't manage to find a solution for it.
Interesting direction, if you'll manage to do it, please update here as well.

**Reply by zdenkolini** [reply]
can't believe this is still unresolved

**Reply by skylupop** [reply]
This is a big issue, in my opinion, and since GIT is actually capable to track the file renaming I can't understand why GitHub cannot.

**Reply by thomasttvo** [reply]
happening to me on Mac

**Reply by tim-phillips** [reply]
This is not an answer! @shirady, could you please unmark as an answer so that this may get the attention it deserves?

**Reply by shirady** [reply]
I unmark it @tim-phillips , I think it was marked by mistake

**Comment by Fred-AtelierSalvetat** [comment]
Hi
I have the same issue:
1. Move files in a commit
-> Github status was correct
2. Modify files content in a second commit
-> Commit appears correctly on GitHib
-> Files Changed" present files as deleted/created.
-> Line counts is then erroneous hence bad metrics :/

**Reply by jabielecki** [reply]
Actually, you can go to GH webUI pull request Files tab and render changes from selected commits (shift+click). The diff of a file renders **up to a certain threshold of changed lines of text**. Over a threshold, the webUI gives up.

1. git mv, content unchanged -> GH shows diff
2. git mv, a few lines of content changed -> GH shows diff
3. git mv, +270/-230 lines changed for a 900 line file -> GH hides diff, but you can still click to show it
4. git mv, more lines than that changed -> GH shows separate create+delete (+900/-900), without a diff

Seen on `GitHub Enterprise Server 3.9.4`

**Reply by yo1995** [reply]
This is also what I observed - it seems GitHub web UI has an implicit changed line limit that determines whether a file is renamed + changed, vs a file is added + deleted. In my case, a +8-6 change https://github.com/Esri/arcgis-maps-sdk-swift-samples/pull/750/changes/748f9a26a059120ee00568dba4a3b85969958e86 was rendered as added + deleted, whereas a +5-5 change https://github.com/Esri/arcgis-maps-sdk-swift-samples/pull/750/changes/a191ea456280921e362a165ba1739fbbd1a84e97 was rendered as renamed + changed.

**Comment by gmnikhil** [comment]
Facing the same issue. `git status` shows `renamed` but github ui showing as `created / deleted`

**Comment by thomasttvo** [comment]
Please fix this issue. This heavily impacts productivity of code review whenever it happen.

**Comment by JanakarM** [comment]
Impacted.

**Reply by JanakarM** [reply]
Can the right person remove marked as answer so that it could be looked after in priority?

**Comment by TamaraFinogina** [comment]
Same here. Git can recognize renaming, but github ui showing the same file as deleted/added

**Comment by Quanalogy** [comment]
We are also hit by this- really what the .... is going on.
On linux doing:

```bash
git mv file file2
git commit -am "rename file to file2"
echo "changes to file 2" >> file2
git commit -am "Changes to file2"
git push
```

This certainly was not an issue on gitlab.

The kicker is if you look at the commit where you rename the file you actually see that it is renamed without changes, looking at the commit where I did the changes to the file for itself also shows this correctly - however in the overview with both commits it suddently is deleted and then created...

@spenserblack any comment on that?

**Reply by spenserblack** [reply]
> however in the overview with both commits it suddently is deleted and then created...

For the sake of this conversation I'll call the commit before the move 1111111, the move commit 2222222, and the commit with file changes 3333333.

GitHub could perhaps be smarter by resolving the diffs between commits individually, that is checking 1111111..2222222 *then* 2222222..3333333 to follow file renames. But this is a result of diffing 1111111..3333333.

I'll try to answer this to be best of my knowledge, but keep in mind that I may be wrong about some things.

There are a few version control systems that track file renames. Git is *not* one of them (at least not completely). Here's a detailed comparison: https://en.wikipedia.org/wiki/Comparison_of_version-control_software#Features

When Git looks at 1111111..2222222, it thinks "I see that when file1 was deleted, and file2 was created with very similar contents, so this is probably a rename." But when Git looks at 1111111..3333333, it's resolving the "total" difference between those two commits, and it thinks "I see that file1 was deleted, and now file2 exists with very different contents, so file1 was deleted and file2 was created."

Here's a quote from the Wikipedia page comparing VCS:

> Git does not explicitly track renames, because by design it does not track individual files. Renames and split of source files are detected after the fact, if the file content does not change dramatically.

[source](https://en.wikipedia.org/wiki/Comparison_of_version-control_software#cite_note-44)

**Comment by saurabhkumar8112** [comment]
This also confused me for a while and then I noticed. git shows correctly but github shows files added and deleted

**Comment by tfedorova** [comment]
I ran into the same issue. My PR has two commits: one with file renames only, and a second with modifications to the renamed files. If you view the changes by commit, the diffs display correctly, but in the “Files changed” tab the files appear as deleted and newly created with different names.

It makes code reviews significantly harder.

**Comment by Shahz-B** [comment]
Same here, I moved a file from one directory to another and committed the change. Both my local Git diff and GitHub’s UI correctly showed it as a rename.

However, after I made a second commit with a bunch of changes to that same file, my local Git still recognizes it as a rename, but GitHub’s PR view now shows it as a Deleted and Created file instead of a renamed file with modifications.

Please consider improving rename detection so GitHub can still track renames even when the file is significantly modified.

**Comment by cederom** [comment]
Have the same issue but shell git 2.39.5 looks like a git bug :-(

The solution is to split move/rename and content change into several commits ;-)

**Reply by vidgeus** [reply]
> Have the same issue but shell git 2.39.5 looks like a git bug :-(
> 
> The solution is to split move/rename and content change into several commits ;-)

you can have any number of commits, even separate commits with just the renames, but in a Pull request you review all changes so the diff shows as if it was a single commit when looking at all changes which has the problem of not following renames.
